### PR TITLE
[Utility] Foundational bugs, tests, code cleanup and improvements (2/3)

### LINKED ISSUE #504: [Utility] Foundational bugs, tests, code cleanup and improvements (2/3)
## Objective

Partial refactor, cleanup, and code improvement of the utility module to set a foundation for M3.

## Origin Document

Follow-up work to #475. Address the items not refactored in #475 as noted on PR #503.

When work on #443 started, while also looking at #325, it became evident that the logic in the utility module is hard to follow and understand while also being filled with minor hard-to-find bugs.

## Goals

- **Unblock development**: Create a foundation on which we design utility protocols and implement new features
- **Reduce cognitive overhead**: Improve code readability and navigation for both internal and external contributors
- **Enable onboarding other engineers**: Clarify both source code and tests through comment and source code improvements

## Deliverable

- [ ] A PR that updates the portions of the utility module noted above
- [ ] Documentation & explanation of the changes that were made and next steps
- [ ] Reviewing and updating the following files where appropriate:
	- [x] ├── application.go
	- [x] ├── application_test.go
	- [x] ├── block.go
	- [ ] ├── block_test.go
	- [x] ├── context.go
	- [ ] ├── doc
	- [ ] │   ├── CHANGELOG.md
	- [ ] │   ├── PROTOCOL_RELAY.md
	- [ ] │   ├── PROTOCOL_SESSION.md
	- [ ] │   └── README.md
	- [ ] ├── gov.go
	- [ ] ├── gov_test.go
	- [ ] ├── message_handler.go
	- [ ] ├── message_test.go
	- [x] ├── module.go
	- [x] ├── module_test.go
	- [ ] ├── service
	- [ ] │   └── service.go
	- [x] ├── session.go
	- [x] ├── transaction.go
	- [ ] ├── transaction_test.go
	- [ ] ├── types
	- [x] │   ├── proto
	- [x] │   │   ├── message.proto
	- [x] │   │   ├── stake_status.proto
	- [x] │   │   ├── transaction.proto
	- [x] │   │   ├── tx_result.proto
	- [x] │   │   └── utility_messages.proto
	- [ ] │   ├── transaction.go
	- [ ] │   ├── transaction_test.go
	- [x] │   ├── tx_result.go
	- [x] └── validator.go

## Non-goals / Non-deliverables

- Updating the entire module in a single PR (PR would be too large)
- Adding new features to the utility module
- Increasing/improving the test coverage to the po

## General issue deliverables

- [ ] Update the appropriate CHANGELOG(s)
- [ ] Update any relevant local/global README(s)
- [ ] Update relevant source code tree explanations
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

## Testing Methodology

- [ ] **All tests**: `make test_all`
- [ ] **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: @jessicadaugherty
**Co-owner**: @Olshansk

### PR DESCRIPTION
**Pull request by Olshansk**
## Description

The second of three changes necessary to refactor and improve the utility module to enable implementation of future, more complex, interfaces.

While the list of changes in #503 was more bug & testing focused, this PR was more readability & documentation focused in preparation for M3.

## Issue

Fixes #504

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [x] Bug fix
- [x] Code health or cleanup
- [x] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Fixed bug where we were not removing txs from the mempool of replicas
- Improved the readability of the `Block` lifecycle and `Message` validation
- Replace unnecessary functions in such as `Store()`, `getStoreAndHeight()` and a few otherwise
- Moved the `Transaction` proto to the shared module
- Added documentation to `TxResult` result and removed `DefaultTxResult` and 
- Added extensive documentation to functions involved in the Block lifecycle
- Removed unnecessary fields from the `BlockHeader`
- Renamed `GenericParam` to `ServiceURL`
- Consolidated `StakeStatus` in the shared package
- Renamed governance parametesr related to rate limitting Applications
- Renamed `applyTx` to `hydrateTx` and added documentation on its functionality

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

**Comment by Olshansk** [comment]
@deblasis I didn't do a super thorough self-review on this yet, and haven't run it with LocalNet yet (unit tests pass), but I think it's ready for review and a good step in the right direction.

File: consensus/hotstuff_leader.go [hunk 1]
```diff
@@ -392,8 +392,14 @@ func (m *consensusModule) prepareAndApplyBlock(qc *typesCons.QuorumCertificate)
 		return nil, err
 	}
 
+	prevHeight := int64(m.height) - 1
+	readCtx, err := m.GetBus().GetPersistenceModule().NewReadContext(prevHeight)
+	if err != nil {
+		return nil, err
+	}
+
 	// IMPROVE: This data can be read via an ephemeral read context - no need to use the utility's persistence context
```
Comment (deblasis) [inline consensus/hotstuff_leader.go:401]:
```suggestion
```

File: runtime/genesis/proto/genesis.proto [hunk 2]
```diff
@@ -21,18 +21,18 @@ message GenesisState {
   Params params = 10;
 }
 
-// DISCUSS(drewskey): Explore a more general purpose "feature flag" like approach for this.
+// TECHDEBT: Explore a more general purpose "feature flag" approach that makes it easy to add/remove
+// parametesr and add activation heights for them as well.
 message Params {
+  reserved 4, 59; // Deprecated
```
Comment (deblasis) [inline runtime/genesis/proto/genesis.proto:27]: TOL: I agree that this is the right approach but since we are "early" perhaps we could just simply change the field indexes instead of reserving the ones that we removed.

It would make, IMHO, things easier to understand to newcomers expecially because there's no history yet around this fields. 
Once we are live/in testnet, of course we should deprecate them this way.

File: runtime/test_artifacts/generator.go [hunk 3]
```diff
@@ -7,7 +7,7 @@ import (
 
 	"github.com/pokt-network/pocket/runtime/configs"
 	"github.com/pokt-network/pocket/runtime/genesis"
-	"github.com/pokt-network/pocket/runtime/test_artifacts/keygenerator"
+	keygen "github.com/pokt-network/pocket/runtime/test_artifacts/keygenerator"
```
Comment (deblasis) [inline runtime/test_artifacts/generator.go:10]: Maybe we can rename the package itself ;)

File: runtime/test_artifacts/generator.go [hunk 4]
```diff
@@ -85,35 +85,30 @@ func NewAccounts(n int, privateKeys ...string) (accounts []*coreTypes.Account) {
 //	types of actors which needs to be fixed.
 func NewActors(actorType coreTypes.ActorType, n int) (actors []*coreTypes.Actor, privateKeys []string) {
 	for i := 0; i < n; i++ {
-		genericParam := getServiceUrl(i + 1)
-		if int32(actorType) == int32(coreTypes.ActorType_ACTOR_TYPE_APP) {
-			genericParam = DefaultMaxRelaysString
-		}
-		actor, pk := NewDefaultActor(int32(actorType), genericParam)
+		serviceUrl := getServiceURL(i + 1)
```
Comment (deblasis) [inline runtime/test_artifacts/generator.go:88]: There's some inconsistency with the `URL` and `Url` suffix here and also I think in the `defaults`, let's pick one and stick to it.

File: shared/core/types/transaction.go [hunk 5]
```diff
@@ -2,48 +2,49 @@ package types
...
 	Bytes() ([]byte, error)
 	Equals(tx2 ITransaction) bool
 }
 
-func (tx *Transaction) ValidateBasic() Error {
+func (tx *Transaction) ValidateBasic() error {
 	// Nonce cannot be empty to avoid transaction replays
 	if tx.Nonce == "" {
-		return ErrEmptyNonce()
+		return fmt.Errorf("nonce cannot be empty")
```
Comment (deblasis) [inline shared/core/types/transaction.go:42]: Any particular reason why you want to move away from the typed error pattern that used the `Error` type?

If we are moving away from it (like it seems given your changes) I guess we should axe it all together otherwise it's just unused code. 🤔

Generally speaking I think we should instead embrace the usage of the pattern across the whole codebase

File: shared/core/types/transaction_test.go [hunk 6]
```diff
@@ -75,47 +59,44 @@ func TestTransaction_ValidateBasic(t *testing.T) {
 	txNoNonce := proto.Clone(&tx).(*Transaction)
 	txNoNonce.Nonce = ""
 	er = txNoNonce.ValidateBasic()
-	require.Equal(t, ErrEmptyNonce().Code(), er.Code())
+	require.Error(t, er)
+	// require.Equal(t, ErrEmptyNonce().Code(), er.Code())
```
Comment (deblasis) [inline shared/core/types/transaction_test.go:63]: Are we moving away from using `Error`? 

More here: https://github.com/pokt-network/pocket/pull/550/files#r1118743922

File: shared/core/types/transaction_test.go [hunk 7]
```diff
@@ -75,47 +59,44 @@ func TestTransaction_ValidateBasic(t *testing.T) {
...
 
 	txInvalidMessageAny := proto.Clone(&tx).(*Transaction)
 	txInvalidMessageAny.Msg = nil
 	er = txInvalidMessageAny.ValidateBasic()
-	require.Equal(t, ErrProtoFromAny(er).Code(), er.Code())
+	require.Error(t, er)
+	// require.Equal(t, ErrProtoFromAny(er).Code(), er.Code())
```
Comment (deblasis) [inline shared/core/types/transaction_test.go:69]: Are we moving away from using `Error`? 

More here: https://github.com/pokt-network/pocket/pull/550/files#r1118743922

File: shared/core/types/transaction_test.go [hunk 8]
```diff
@@ -75,47 +59,44 @@ func TestTransaction_ValidateBasic(t *testing.T) {
...
-	require.Equal(t, ErrEmptyPublicKey().Code(), er.Code())
+	require.Error(t, er)
+	// require.Equal(t, ErrEmptyPublicKey().Code(), er.Code())
 
 	txInvalidPublicKey := proto.Clone(&tx).(*Transaction)
 	txInvalidPublicKey.Signature.PublicKey = []byte("publickey")
 	err = txInvalidPublicKey.ValidateBasic()
-	require.Equal(t, ErrNewPublicKeyFromBytes(err).Code(), err.Code())
+	require.Error(t, er)
+	// require.Equal(t, ErrNewPublicKeyFromBytes(err).Code(), err.Code())
```
Comment (deblasis) [inline shared/core/types/transaction_test.go:87]: Are we moving away from using `Error`? 

More here: https://github.com/pokt-network/pocket/pull/550/files#r1118743922

File: shared/core/types/transaction_test.go [hunk 9]
```diff
@@ -75,47 +59,44 @@ func TestTransaction_ValidateBasic(t *testing.T) {
...
-}
-
-func newTestingMsgSend(_ *testing.T) Message {
-	return &MessageSend{
-		FromAddress: testingSenderAddr,
-		ToAddress:   testingToAddr,
-		Amount:      defaultAmount,
-	}
+	require.Error(t, er)
+	// require.Equal(t, ErrSignatureVerificationFailed().Code(), er.Code())
```
Comment (deblasis) [inline shared/core/types/transaction_test.go:94]: Are we moving away from using `Error`? 

More here: https://github.com/pokt-network/pocket/pull/550/files#r1118743922

File: utility/validator.go [hunk 10]
```diff
@@ -1,24 +1,77 @@
...
+
+// burnValidator burns a validator's stake based on governance parameters for missing blocks
+// and begins unstaking if the stake falls below the necessary threshold
+// REFACTOR: Extend this to support burning other actors types & pools once the logic is implemented
 // ADDTEST: There are no good tests for this functionality, which MUST be added.
-func (u *utilityContext) burnValidator(burnPercent int, addr []byte) typesUtil.Error {
-	// TODO: Will need to extend this to support burning from other actors types & pools when the logic is implemented
-	validatorActorType := coreTypes.ActorType_ACTOR_TYPE_VAL
-	validatorPool := coreTypes.Pools_POOLS_VALIDATOR_STAKE
+func (u *utilityContext) burnValidator(addr []byte) typesUtil.Error {
```
Comment (deblasis) [inline utility/validator.go:60]: Confusingly, IMHO, this returns `Error` but it has been removed elsewhere

Related: https://github.com/pokt-network/pocket/pull/550/files#r1118743922

File: utility/transaction_test.go [hunk 11]
```diff
@@ -19,26 +20,26 @@ var (
...
+	// require.NoError(t, err)
+	// require.Equal(t, []byte(signer.Address().String()), msg.GetSigner())
+	// feeBig, err := ctx.getMessageSendFee()
+	// require.NoError(t, err)
+
+	// expectedAfterBalance := big.NewInt(0).Sub(startingBalance, feeBig)
+	// amount, err := ctx.getAccountAmount(signer.Address())
+	// require.NoError(t, err)
+	// require.Equal(t, expectedAfterBalance, amount, "unexpected after balance")
 }
```
Comment (deblasis) [inline utility/transaction_test.go:36]: I am not sure I understand what you are trying to do here.

File: utility/service/service.go [hunk 12]
```diff
@@ -119,17 +119,17 @@ func (r *relay) ReapStoreForHashCollision(sessionBlockHeight int64, hashEndWith
...
-	// a proper verification of the volume completed. Send volumeRelays to fishermanServiceURL
+	// a proper verification of the volume completed. Send volumeRelays to fishermanServiceUrl
 	// through http.
 
 	// NOTE: an alternative design is a 2 step, claim - proof lifecycle where the individual servicers
 	// build a merkle sum index tree from all the relays, submits a root and subsequent merkle proof to the
 	// network.
 	//
-	// Pros: Can report volume metrics directly to the chain in a trustless fashion
+	// Pros:cs directly to the Can report volume metri chain in a trustless fashion
```
Comment (deblasis) [inline utility/service/service.go:132]:
```suggestion
	// Pros: Can report volume metrics directly to the chain in a trustless fashion
```

File: shared/core/types/proto/actor.proto [hunk 13]
```diff
@@ -12,12 +12,19 @@ enum ActorType {
...
+}
+
 message Actor {
   ActorType actor_type = 1;
   string address = 2;
   string public_key = 3;
-  repeated string chains = 4;
-  string generic_param = 5; // TODO(#313): Generalize `GenericParam` to a different name
+  repeated string chains = 4; // Not applicable to `Validator` actors
+  string service_url = 5; // Not applicable to `Application` actors
```
Comment (deblasis) [inline shared/core/types/proto/actor.proto:27]: I think it's a bit smelly that we have fields that are only applicable to some actors in here.

Especially now that `generic_param` was promoted to `service_url`

Not actionable for this PR of course.

I would perhaps close or modify #313 and leave a TODO here. Wdyt? I can help if needed

File: utility/doc/CHANGELOG.md [hunk 14]
```diff
@@ -7,6 +7,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- Added thorough documentation in different parts of the codebase
+- Replace the `Store()` function with a `.store` accessor
+- Removed the unnecessary `getStoreAndHeight()` function
+- Simplified `calculateMaxAppRelays` to a temporary `calculateAppSessionTokens` implementation
+- Improved the readability of the `Block` lifecycle and `Message` validation
+- Simplified the testing utilities used
+- Updated the `session` interface
+- Renamed `applyTx` to `hydrateTx` and added documentation on its functionality
+- Removed `DefaultTxResult` and added documentation to `TxResult` result
+-
```
Comment (deblasis) [inline utility/doc/CHANGELOG.md:21]:
```suggestion
```

File: shared/node.go [hunk 15]
```diff
@@ -126,6 +126,7 @@ func (m *Node) GetBus() modules.Bus {
 	return m.bus
 }
 
+// TECHDEBT: The `shared` package has dependencies on types in the individual modules.
```
Comment (deblasis) [inline shared/node.go:129]: N.A.: Right, I guess that's why I added messages in `messaging` instead of the various packages.

File: utility/types/error.go [hunk 16]
```diff
@@ -38,6 +39,8 @@ func NewError(code Code, msg string) Error {
 	}
 }
 
+// NextCode: 133
+// CONSIDERATION: Should these be a proto enum or
```
Comment (deblasis) [inline utility/types/error.go:43]: or? :)

File: utility/types/error.go [hunk 17]
```diff
@@ -38,6 +39,8 @@ func NewError(code Code, msg string) Error {
 	}
 }
 
+// NextCode: 133
```
Comment (deblasis) [inline utility/types/error.go:42]: nice

File: utility/types/error.go [hunk 18]
```diff
@@ -325,7 +330,7 @@ func ErrUnequalRounds() Error {
 	return NewError(CodeUnequalRoundsError, UnequalRoundsError)
 }
 
-func ErrInvalidServiceUrl(reason string) Error {
+func ErrInvalidServiceURL(reason string) Error {
```
Comment (deblasis) [inline utility/types/error.go:333]: Another example of inconsistency around `URL` / `Url`. I recommend a case-sensitive search and renaming consistently across the board.

File: utility/types/message.go [hunk 19]
```diff
@@ -180,6 +185,7 @@ func validateRelayChains(chains []string) Error {
 }
 
 func getMessageType(msg Message) string {
+	// GetMessageName() is a method on the Message interface, so we can't use it here.
```
Comment (deblasis) [inline utility/types/message.go:188]: If you really think this comment is necessary, I'd suggest renaming this `getMessageType` to something like `getProtoDescriptorName` so that the intent is clear

File: utility/transaction.go [hunk 20]
```diff
@@ -40,57 +42,78 @@ func (u *utilityModule) CheckTransaction(txProtoBytes []byte) error {
...
+// the relevant data from the embedded signed Message.
+func (u *utilityContext) hydrateTx(tx *coreTypes.Transaction, index int) (modules.TxResult, typesUtil.Error) {
+	msg, err := u.anteHandleMessage(tx)
 	if err != nil {
 		return nil, err
 	}
-	return tx.ToTxResult(u.height, index, signer, msg.GetMessageRecipient(), msg.GetMessageName(), u.handleMessage(msg))
+	msgHandlingResult := u.handleMessage(msg)
+	// INCOMPLETE:
+	return typesUtil.TxToTxResult(tx, u.height, index, msg, msgHandlingResult)
```
Comment (deblasis) [inline utility/transaction.go:54]: Curious: do you see `TxResult` as part of `coreTypes` or `typesUtil`?

File: utility/transaction.go [hunk 21]
```diff
@@ -40,57 +42,78 @@ func (u *utilityModule) CheckTransaction(txProtoBytes []byte) error {
...
 }
 
-func (u *utilityContext) anteHandleMessage(tx *typesUtil.Transaction) (msg typesUtil.Message, signer string, err typesUtil.Error) {
-	msg, err = tx.GetMessage()
-	if err != nil {
-		return nil, "", err
+// anteHandleMessage handles basic validation of the message in the Transaction before it is processed
+// REFACTOR: Splitting this into a `feeValidation`, `signerValidation`, and `messageValidation` etc
+// would make it more modular and readable.
```
Comment (deblasis) [inline utility/transaction.go:59]: 💯

File: utility/module_test.go [hunk 22]
```diff
@@ -73,35 +71,52 @@ func newTestingUtilityContext(t *testing.T, height int64) *utilityContext {
...
-		savePointsSet:      make(map[string]struct{}),
-		savePointsList:     make([][]byte, 0),
+	ctx := &utilityContext{
+		logger:         logger.Global.CreateLoggerForModule(modules.UtilityModuleName),
+		height:         height,
+		store:          persistenceContext,
+		savePointsSet:  make(map[string]struct{}),
+		savePointsList: make([][]byte, 0),
+	}
+	ctx.IntegratableModule.SetBus(testUtilityMod.GetBus())
```
Comment (deblasis) [inline utility/module_test.go:81]:
```suggestion
	ctx.SetBus(testUtilityMod.GetBus())
```

File: utility/block.go [hunk 23]
```diff
@@ -99,9 +99,11 @@ func (u *utilityContext) ApplyBlock() (string, error) {
 		return "", err
 	}
 
+	// mempool := u.GetBus().GetUtilityModule().GetMempool()
+
```
Comment (deblasis) [inline utility/block.go:103]:
```suggestion
```

File: utility/block.go [hunk 24]
```diff
@@ -113,18 +115,18 @@ func (u *utilityContext) ApplyBlock() (string, error) {
...
 
-		if err := u.persistenceContext.IndexTransaction(txResult); err != nil {
+		if err := u.store.IndexTransaction(txResult); err != nil {
 			u.logger.Fatal().Err(err).Msgf("TODO(#327): We can apply the transaction but not index it. Crash the process for now: %v\n", err)
 		}
 
 		// TODO: if found, remove transaction from mempool.
 		// DISCUSS: What if the context is rolled back or cancelled. Do we add it back to the mempool?
-		// if err := u.Mempool.RemoveTx(tx); err != nil {
+		// if err := mempool.RemoveTx(tx.Bytes()); err != nil {
```
Comment (deblasis) [inline utility/block.go:129]:
```suggestion
		// if err := u.GetBus().GetUtilityModule().GetMempool().RemoveTx(tx.Bytes()); err != nil {
```

File: utility/application_test.go [hunk 25]
```diff
@@ -10,7 +10,9 @@ import (
 func TestUtilityContext_CalculateMaxAppRelays(t *testing.T) {
 	ctx := newTestingUtilityContext(t, 1)
 	actor := getFirstActor(t, ctx, coreTypes.ActorType_ACTOR_TYPE_APP)
-	newMaxRelays, err := ctx.calculateMaxAppRelays(actor.GetStakedAmount())
+	appSessionTokens, err := ctx.calculateAppSessionTokens(actor.StakedAmount)
 	require.NoError(t, err)
-	require.Equal(t, actor.GetGenericParam(), newMaxRelays)
+	// TODO: These are hardcoded values based on params from the genesis file. Expand on tests
+	// when implementing the Application protocol.
+	require.Equal(t, "100000000000000", appSessionTokens)
```
Comment (deblasis) [inline utility/application_test.go:17]: I could ask you to add a constant for this but I won't 🙃

do_it_palpatine.gif

**Review by deblasis — CHANGES_REQUESTED** [review]
Yo @Olshansk! I gave this a first pass. Solid work as usual. 

Some bits and bolts and Qs. Depending on the priority of this vs what I am working on currently let me know if and how I can help you here. 🫡

**Review by Olshansk — COMMENTED** [review]
@deblasis Thank you for the quick turnaround on the review. 

PTAL at the changes I made along with the comments I left on the unresolved comments. 

I think you have a good sense of some future follow-up work from the changes + the couple of issues I created out of it.

I want to prioritize getting this in given the size of the PR and because I'll be OOO later this week so there won't be much time to maintain & rebase it.

**Review by deblasis — APPROVED** [review]
🚢 🇮🇹

**Comment by Olshansk** [comment]
@deblasis Thank you again for the quick turnaround. I know this was a bit rushed but I did a local rerun of the unit tests and LocalNet (with docker-compose + sending a transaction) and verified that it works.

> 🚢 🇮🇹

I'm on season 2 of the White Lotus, so I guess I'll see you in Sicily.

**Comment by deblasis** [comment]
> @deblasis Thank you again for the quick turnaround. I know this was a bit rushed but I did a local rerun of the unit tests and LocalNet (with docker-compose + sending a transaction) and verified that it works.

👍💯

> > 🚢 🇮🇹
> 
> I'm on season 2 of the White Lotus, so I guess I'll see you in Sicily.

Never watched, no idea what's about but I hear it's pop culture already. 

LOL. Funny. I might be going to Sicily for real in a few weeks... 
MakeYouAnOfferGIF

**Comment by jessicadaugherty** [comment]
Part 3/3 can be found here: https://github.com/pokt-network/pocket/issues/574
